### ISSUE TITLE
Model.update should respect the key order of condition

### ISSUE THREAD

**Issue by jasson15** [Automattic/mongoose#3317]
In lib/model.js, `Model.update` changes the order of condition. It seems not a big deal, but it matters when coming along with `upsert`.

For example, without specifying `retainKeyOrder`, it is entirely possible that:

``` javascript
Model.update({
  sub: {
    sub_id: '12345',
    sub_cat: 'automotive',
  }
}, {
  $set: { count: 1 },
}, { upsert: true });
```

But results in an altered document:

``` json
{ "sub": { "sub_cat": "automotive", "sub_id": "12345"} , "count": 1 }
```

In this case, we won't be able to find this document via the following because **the order matters**:

``` javascript
Model.find({sub: {sub_id: '123456', sub_cat: 'automotive'}})
```

I suggest to fix [Line1758](https://github.com/Automattic/mongoose/blame/master/lib/model.js#L1758) as below:

``` javascript
conditions = utils.clone(conditions, { retainKeyOrder: true });
```

**Comment by jasson15** [comment]
I tried by this (according to d098ad3):

```javascript
conditions = utils.clone(conditions, this.schema.options.toObject); 
```

I wrote a test case for this:

1. `upsert` a document
2. `findOne`
3. compare their subdocument

But my test case fails because operations still not respect to this option, such as `Model.findOne` which crashes my test case.

The test case:

```javascript
it('query should keep order even if upsert', function(done) {
  var mydoc = {
    sub: {
      sub_id: '1234',
      sub_cat: 'abcd',
    }
  };
  MyModel.update(mydoc, {}, { upsert: true }, function(err) {
    assert.ifError(err);
    MyModel.findOne({}, function(err, doc) {
      assert.ifError(err);
      assert.equal(JSON.stringify(doc.sub)), JSON.stringify(mydoc.sub));
    done();
  });
});
```

**Comment by vkarpov15** [comment]
I believe this was fixed in 4.1.4 re: #3215. However, I agree 100% that bson key order does matter, which is the reason for #2749 and always using natural key order instead of mongoose's legacy behavior of "reverse everything because its marginally faster." Unfortunately MongoDB has never been explicitly clear on whether or not bson key order matters, which is why mongoose assumed that it didn't when it was first written.

**Comment by vkarpov15** [comment]
This should be fixed in >= 4.6.0
